File: docs/source/_static/examples/cpp/custom_plugin.cpp
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See LICENSE.txt in the project root for
 * license information.
 */

// Custom plugin examples for QDK/Chemistry.
//
// This file demonstrates how to extend QDK/Chemistry with custom plugins:
// 1. Adding a new backend for an existing algorithm type
// 2. Defining an entirely new algorithm type

#include <qdk/chemistry.hpp>

// -----------------------------------------------------------------------------
// start-cell-custom-settings
class CustomScfSettings
    : public qdk::chemistry::algorithms::ElectronicStructureSettings {
 public:
  CustomScfSettings() : ElectronicStructureSettings() {
    // Define additional settings beyond the inherited defaults
    set_default("custom_option", "default_value");
  }
};
// end-cell-custom-settings
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-custom-scf-solver
#include <qdk/chemistry/algorithms/scf.hpp>

#include "external_chemistry_package.hpp"

class CustomScfSolver : public qdk::chemistry::algorithms::ScfSolver {
 public:
  CustomScfSolver() { _settings = std::make_unique<CustomScfSettings>(); }

  std::string name() const override { return "custom"; }

 protected:
  std::pair<double, std::shared_ptr<qdk::chemistry::data::Wavefunction>>
  _run_impl(std::shared_ptr<qdk::chemistry::data::Structure> structure,
            int charge, int spin_multiplicity,
            std::optional<std::shared_ptr<qdk::chemistry::data::Orbitals>>
                initial_guess) override {
    // Convert to external format
    auto external_mol = convert_to_external_format(structure);

    // Execute external calculation
    auto basis = _settings->get<std::string>("basis_set");
    auto [energy, external_orbitals] =
        external_package::run_scf(external_mol, basis);

    // Convert results to QDK format
    auto wavefunction = convert_to_qdk_wavefunction(external_orbitals);

    return {energy, wavefunction};
  }
};
// end-cell-custom-scf-solver
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-registration
#include <qdk/chemistry/algorithms/scf.hpp>

// Static registration during library initialization
static auto registration =
    qdk::chemistry::algorithms::ScfSolver::register_implementation(
        []() { return std::make_unique<CustomScfSolver>(); });
// end-cell-registration
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-descriptor-settings
class MolecularDescriptorSettings : public qdk::chemistry::data::Settings {
 public:
  MolecularDescriptorSettings() {
    set_default<bool>("normalize", false, "Normalize the descriptor");
  }
};
// end-cell-descriptor-settings
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-descriptor-base-class
class MolecularDescriptorCalculator
    : public qdk::chemistry::algorithms::Algorithm<
          MolecularDescriptorCalculator,
          double,                                            // Return type
          std::shared_ptr<qdk::chemistry::data::Structure>>  // Input type
{
 public:
  std::string type_name() const final {
    return "molecular_descriptor_calculator";
  }
};
// end-cell-descriptor-base-class
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-descriptor-factory
struct MolecularDescriptorCalculatorFactory
    : public qdk::chemistry::algorithms::AlgorithmFactory<
          MolecularDescriptorCalculator, MolecularDescriptorCalculatorFactory> {
  static std::string algorithm_type_name() {
    return "molecular_descriptor_calculator";
  }

  static std::string default_algorithm_name() { return "nuclear_charge"; }
};
// end-cell-descriptor-factory
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-descriptor-implementations
class NuclearChargeDescriptor : public MolecularDescriptorCalculator {
 public:
  NuclearChargeDescriptor() {
    _settings = std::make_unique<MolecularDescriptorSettings>();
  }

  std::string name() const override { return "nuclear_charge"; }

 protected:
  double _run_impl(std::shared_ptr<qdk::chemistry::data::Structure> structure)
      const override {
    const auto& charges = structure->get_nuclear_charges();
    double descriptor = 0.0;
    for (double charge : charges) descriptor += charge;
    if (_settings->get<bool>("normalize") && structure->get_num_atoms() > 0) {
      descriptor /= static_cast<double>(structure->get_num_atoms());
    }
    return descriptor;
  }
};
// end-cell-descriptor-implementations
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-descriptor-registration
static auto descriptor_registration = []() {
  MolecularDescriptorCalculatorFactory::register_instance(
      []() { return std::make_unique<NuclearChargeDescriptor>(); });
  return true;
}();
// end-cell-descriptor-registration
// -----------------------------------------------------------------------------

// -----------------------------------------------------------------------------
// start-cell-descriptor-usage
#include <iostream>
#include <qdk/chemistry.hpp>

using namespace qdk::chemistry::algorithms;

int main() {
  // After registration, use like any built-in algorithm
  auto calculator =
      MolecularDescriptorCalculatorFactory::create("nuclear_charge");
  calculator->settings().set("normalize", true);

  // List available implementations
  auto available = MolecularDescriptorCalculatorFactory::available();
  std::cout << "Available molecular descriptor calculators: ";
  for (const auto& name : available) {
    std::cout << name << " ";
  }
  std::cout << std::endl;

  return 0;
}
// end-cell-descriptor-usage
// -----------------------------------------------------------------------------

File: docs/source/_static/examples/python/custom_plugin.py
"""Custom plugin examples for QDK/Chemistry.

This module demonstrates how to extend QDK/Chemistry with custom plugins:
1. Adding a new backend for an existing algorithm type
2. Defining an entirely new algorithm type
"""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

################################################################################
# start-cell-custom-settings
from qdk_chemistry.data import ElectronicStructureSettings


class CustomScfSettings(ElectronicStructureSettings):
    """Settings for the custom SCF solver."""

    def __init__(self):
        super().__init__()
        # Define additional settings beyond the inherited defaults
        self._set_default(
            "custom_option",
            "string",
            "default_value",
            "Description of the custom option",
        )


# end-cell-custom-settings
################################################################################

################################################################################
# start-cell-custom-scf-solver
from qdk_chemistry.algorithms import ScfSolver
from qdk_chemistry.data import (
    BasisSet,
    Orbitals,
    Structure,
    Wavefunction,
)


class CustomScfSolver(ScfSolver):
    """Custom SCF solver wrapping an external chemistry package."""

    def __init__(self):
        super().__init__()
        self._settings = CustomScfSettings()

    def name(self) -> str:
        return "custom"

    def _run_impl(
        self,
        structure: Structure,
        charge: int,
        spin_multiplicity: int,
        basis_or_guess: Orbitals | BasisSet | str | None = None,
    ) -> tuple[float, Wavefunction]:
        """Perform a self-consistent field (SCF) calculation using a custom backend.

        This method should convert the input structure to the external format, run the SCF calculation
        using the specified method and basis set, and return the electronic energy and wavefunction
        in QDK/Chemistry format.

        Args:
            structure: The molecular structure to be calculated.
            charge: The total charge of the molecular system.
            spin_multiplicity: The spin multiplicity (2S+1) of the system.
            basis_or_guess: Basis set information or initial guess, which can be:
                - An Orbitals object (used as initial guess)
                - A BasisSet object
                - A string specifying the basis set name
                - None (use default from settings)

        Returns:
            Tuple of (energy, wavefunction)
        """
        # Convert to external format

        # Execute external calculation

        # Convert results to QDK format

        # energy = 0.0
        # wavefunction = Wavefunction(...)
        # return energy, wavefunction
        return 0.0, None


# end-cell-custom-scf-solver
################################################################################

################################################################################
# start-cell-registration
from qdk_chemistry.algorithms.registry import register

# Registration during module import
register(lambda: CustomScfSolver())
# end-cell-registration
################################################################################

################################################################################
# start-cell-usage-after-registration
from qdk_chemistry.algorithms import available, create
from qdk_chemistry.data import Structure

# Define a molecular structure (e.g., H2 molecule)
coords = [[0.0, 0.0, 0.0], [0.0, 0.0, 1.4]]
molecule = Structure(coords, symbols=["H", "H"])

# Instantiate the custom solver
solver = create("scf_solver", "custom")
# energy, wavefunction = solver.run(
#     molecule, charge=0, spin_multiplicity=1, basis_or_guess="sto-3g"
# )

# Verify registration
print(available("scf_solver"))  # [..., 'custom']
# end-cell-usage-after-registration
################################################################################

################################################################################
# start-cell-descriptor-settings
from qdk_chemistry.data import Settings


class MolecularDescriptorSettings(Settings):
    """Settings for molecular descriptor algorithms."""

    def __init__(self):
        super().__init__()
        self._set_default("normalize", "bool", False, "Normalize the descriptor")


# end-cell-descriptor-settings
################################################################################

################################################################################
# start-cell-descriptor-base-class
from qdk_chemistry.algorithms.base import Algorithm


class MolecularDescriptorCalculator(Algorithm):
    """Abstract base class for molecular descriptor algorithms."""

    def type_name(self) -> str:
        return "molecular_descriptor_calculator"


# end-cell-descriptor-base-class
################################################################################

################################################################################
# start-cell-descriptor-factory
from qdk_chemistry.algorithms.base import AlgorithmFactory


class MolecularDescriptorCalculatorFactory(AlgorithmFactory):
    """Factory for creating molecular descriptor calculators."""

    def algorithm_type_name(self) -> str:
        return "molecular_descriptor_calculator"

    def default_algorithm_name(self) -> str:
        return "nuclear_charge"


# end-cell-descriptor-factory
################################################################################

################################################################################
# start-cell-descriptor-implementations
from qdk_chemistry.data import Structure


class NuclearChargeDescriptor(MolecularDescriptorCalculator):
    """Calculator for a nuclear-charge molecular descriptor."""

    def __init__(self):
        super().__init__()
        self._settings = MolecularDescriptorSettings()

    def name(self) -> str:
        return "nuclear_charge"

    def _run_impl(self, structure: Structure) -> float:
        descriptor = float(sum(structure.get_nuclear_charges()))
        if self.settings().get("normalize") and structure.get_num_atoms() > 0:
            descriptor /= structure.get_num_atoms()
        return descriptor


# end-cell-descriptor-implementations
################################################################################


################################################################################
# start-cell-mass-descriptor
class MassDescriptor(MolecularDescriptorCalculator):
    """Calculator for a molecular-mass descriptor."""

    def __init__(self):
        super().__init__()
        self._settings = MolecularDescriptorSettings()

    def name(self) -> str:
        return "mass"

    def _run_impl(self, structure: Structure) -> float:
        descriptor = float(sum(structure.get_masses()))
        if self.settings().get("normalize") and structure.get_num_atoms() > 0:
            descriptor /= structure.get_num_atoms()
        return descriptor


# end-cell-mass-descriptor
################################################################################

################################################################################
# start-cell-descriptor-registration
from qdk_chemistry import algorithms

# Register the factory
algorithms.registry.register_factory(MolecularDescriptorCalculatorFactory())

# Register implementations
algorithms.register(lambda: NuclearChargeDescriptor())
algorithms.register(lambda: MassDescriptor())
# end-cell-descriptor-registration
################################################################################

################################################################################
# start-cell-descriptor-usage
from qdk_chemistry.algorithms import available, create

# List available implementations
print(available("molecular_descriptor_calculator"))  # ['mass', 'nuclear_charge']

# Instantiate and configure
calculator = create("molecular_descriptor_calculator", "nuclear_charge")
calculator.settings().set("normalize", True)

# Execute
# descriptor = calculator.run(molecule)
# end-cell-descriptor-usage
################################################################################
